### DISCUSSION TITLE
Изменить font для u8g2

### DISCUSSION THREAD

**Discussion by khvalera** [Spirik/GEM#47] [Q&A] [answered]
Есть ли возможность в библиотеке для u8g2 изменить font, кроме как править GEM_FONT_BIG,  GEM_FONT_SMALL,GEM_FONT_BIG_CYR и GEM_FONT_SMALL_CYR в GEM_u8g2.h?

**Comment by Spirik** [comment]
Привет, @khvalera !

В данный момент поменять шрифт можно только через правку `GEM_u8g2.h`.

**Reply by khvalera** [reply]
Можно как-то в loop() определить выбранный пункт из меню?

**Reply by Spirik** [reply]
> Можно как-то в loop() определить выбранный пункт из меню?

В текущей версии (1.3.4) без модификации исходников этого сделать нельзя. Однако в планах на будущее представить механизм получения подробной информации о текущем состоянии меню через некий объект [state](https://github.com/Spirik/GEM/pull/42#issuecomment-956334863).

В актуальной же версии необходимая для определения текущего пункта меню информация хранится в private полях:
- в `GEM.h`, `GEM_u8g2.h`, `GEM_adafruit_gfx.h` в поле `GEMPage* _menuPageCurrent` (указатель на текущую страницу меню);
- в `GEMPage.h` в поле `byte currentItemNum` (индекс текущего пункта меня) и в методе `GEMItem* getCurrentMenuItem();` (получить указатель на текущий пункт меню).

Если поменять область видимости этих полей и методов с private на public, то, в теории, к ним можно будет получить доступ из пользовательского скетча, что-то вроде: `myMenu._menuPageCurrent->currentItemNum` и `GEMItem* currentItem = myMenu._menuPageCurrent->getCurrentMenuItem()`.

**Reply by khvalera** [reply]
> >
> В текущей версии (1.3.4) без модификации исходников этого сделать нельзя. Однако в планах на будущее представить механизм получения подробной информации о текущем состоянии меню через некий объект [state](https://github.com/Spirik/GEM/pull/42#issuecomment-956334863).

Спасибо за развернутый ответ автору! Для большей гибкости библиотеки было бы неплохо.

**Reply by Spirik** [reply]
Методы для получения текущего пункта меню появились в версии [1.5.0](https://github.com/Spirik/GEM/releases/tag/1.5.0):

* Получение указателя на текущую страницу и пункт меню при помощи методов `::getCurrentMenuPage()`, `GEMPage::getCurrentMenuItem()`, `GEMPage::getCurrentMenuItemIndex()`;
* Поиск пунктов меню из скетча при помощи методов `GEMPage::getMenuItem()` и `GEMItem::getMenuItemNext()`.

**Comment by Spirik** [comment, marked as answer]
Поднимаю тред! В версии [1.5.1](https://github.com/Spirik/GEM/releases/tag/1.5.1) наконец добавилась поддержка кастомных шрифтов:

* Методы `::setFontBig()` и `::setFontSmall()` U8g2 и Adafruit GFX версий GEM для задания "большого" и "маленького" шрифта;
* Поддержка UTF8 шрифтов в U8g2 версии GEM с использованием метода `GEM_u8g2::enableUTF8()` (только для заголовков страниц и пунктов меню - редактирование строк с UTF8 символами в них не поддерживается).

См. подробности в соотвествующем разделе [wiki](https://github.com/Spirik/GEM/wiki/Menu-appearance#changing-fonts).
